[BUG] 'Browser Agent' template creation broken in local Studio

### Describe the Bug

After installing the minimal Mastra Studio canvas through the cli, attempting to install the 'Browser Agent' template to my project resulted in a fatal crash, with a generic ErrorBoundary from the UI framework that had no custom functionality for that particular route. Below is a screenshot that shows the error message, and I will include the full-text below that.

<img width="1904" height="1049" alt="Image" src="https://github.com/user-attachments/assets/f032b17d-023f-4b9e-94d1-b5d2d3783519" />

```
## Unexpected Application Error!

http://localhost:4111/templates/browsing-agent?runId=d59d558c-4c6e-47cb-952b-6b39bdee4e7b&resume=true&provider=openai

Bug 2 - Workflow condition qualityScore >= 6 doesn't re-render when changing the code to 7 isntead
### e?.includes is not a function

TypeError: e?.includes is not a function
    at kln (http://localhost:4111/assets/main-DHJmLzGx.js:585:39011)
    at e9 (http://localhost:4111/assets/main-DHJmLzGx.js:9:48989)
    at x9 (http://localhost:4111/assets/main-DHJmLzGx.js:9:72055)
    at Vme (http://localhost:4111/assets/main-DHJmLzGx.js:9:82551)
    at vge (http://localhost:4111/assets/main-DHJmLzGx.js:9:118645)
    at Bbt (http://localhost:4111/assets/main-DHJmLzGx.js:9:117664)
    at U9 (http://localhost:4111/assets/main-DHJmLzGx.js:9:117491)
    at fge (http://localhost:4111/assets/main-DHJmLzGx.js:9:114228)
    at jge (http://localhost:4111/assets/main-DHJmLzGx.js:9:126416)
    at f_ (http://localhost:4111/assets/main-DHJmLzGx.js:9:124942)

💿 Hey developer 👋

You can provide a way better UX than this when your app throws errors by providing your own `ErrorBoundary` or `errorElement` prop on your route.
```

There were no smoking guns in the Network tabs (500s, etc) but here are the npm dependencies for the local studio: and finally the Console logs:
```
{
  "name": "news-agent",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "dev": "mastra dev",
    "build": "mastra build",
    "start": "mastra start"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "type": "module",
  "engines": {
    "node": ">=22.13.0"
  },
  "dependencies": {
    "@browserbasehq/sdk": "^2.7.0",
    "@mastra/core": "^1.10.0",
    "@mastra/evals": "^1.1.2",
    "@mastra/libsql": "^1.6.4",
    "@mastra/loggers": "^1.0.2",
    "@mastra/memory": "^1.6.1",
    "@mastra/observability": "^1.3.1",
    "@tavily/core": "^0.7.2",
    "exa-js": "^2.8.0",
    "p-limit": "^7.3.0",
    "puppeteer-core": "^24.39.0",
    "zod": "^4.3.6"
  },
  "devDependencies": {
    "@types/node": "^25.4.0",
    "mastra": "^1.3.7",
    "typescript": "^5.9.3"
  }
}
```

Console logs - 

```
Checking completed run after hot reload: {runId: 'ad2cf43a-4d3c-4f76-9f70-eeeaef0b1535'}
main-DHJmLzGx.js:615 🔄 Checking completed run after hot reload: {runId: 'ad2cf43a-4d3c-4f76-9f70-eeeaef0b1535'}
main-DHJmLzGx.js:579 Model updated: {message: 'Model updated successfully'}
main-DHJmLzGx.js:615 🔄 Checking completed run after hot reload: {runId: 'd59d558c-4c6e-47cb-952b-6b39bdee4e7b'}
main-DHJmLzGx.js:615 🔄 Checking completed run after hot reload: {runId: 'd59d558c-4c6e-47cb-952b-6b39bdee4e7b'}
installHook.js:1 Error handled by React Router default ErrorBoundary: TypeError: e?.includes is not a function
    at kln (main-DHJmLzGx.js:585:39011)
    at e9 (main-DHJmLzGx.js:9:48989)
    at x9 (main-DHJmLzGx.js:9:72055)
    at Vme (main-DHJmLzGx.js:9:82551)
    at vge (main-DHJmLzGx.js:9:118645)
    at Bbt (main-DHJmLzGx.js:9:117664)
    at U9 (main-DHJmLzGx.js:9:117491)
    at fge (main-DHJmLzGx.js:9:113874)
    at jge (main-DHJmLzGx.js:9:126416)
    at f_ (main-DHJmLzGx.js:9:124942)
overrideMethod @ installHook.js:1
Wfn @ main-DHJmLzGx.js:611
e9 @ main-DHJmLzGx.js:9
x9 @ main-DHJmLzGx.js:9
Vme @ main-DHJmLzGx.js:9
vge @ main-DHJmLzGx.js:9
Bbt @ main-DHJmLzGx.js:9
U9 @ main-DHJmLzGx.js:9
fge @ main-DHJmLzGx.js:9
jge @ main-DHJmLzGx.js:9
f_ @ main-DHJmLzGx.js:9
Oge @ main-DHJmLzGx.js:9
wge @ main-DHJmLzGx.js:9
hge @ main-DHJmLzGx.js:9
fge @ main-DHJmLzGx.js:9
Ige @ main-DHJmLzGx.js:9
R @ main-DHJmLzGx.js:2Understand this error
installHook.js:1 Error handled by React Router default ErrorBoundary: TypeError: e?.includes is not a function
    at kln (main-DHJmLzGx.js:585:39011)
    at e9 (main-DHJmLzGx.js:9:48989)
    at x9 (main-DHJmLzGx.js:9:72055)
    at Vme (main-DHJmLzGx.js:9:82551)
    at vge (main-DHJmLzGx.js:9:118645)
    at Bbt (main-DHJmLzGx.js:9:117664)
    at U9 (main-DHJmLzGx.js:9:117491)
    at fge (main-DHJmLzGx.js:9:114228)
    at jge (main-DHJmLzGx.js:9:126416)
    at f_ (main-DHJmLzGx.js:9:124942)
overrideMethod @ installHook.js:1
Wfn @ main-DHJmLzGx.js:611
e9 @ main-DHJmLzGx.js:9
x9 @ main-DHJmLzGx.js:9
Vme @ main-DHJmLzGx.js:9
vge @ main-DHJmLzGx.js:9
Bbt @ main-DHJmLzGx.js:9
U9 @ main-DHJmLzGx.js:9
fge @ main-DHJmLzGx.js:9
jge @ main-DHJmLzGx.js:9
f_ @ main-DHJmLzGx.js:9
Oge @ main-DHJmLzGx.js:9
wge @ main-DHJmLzGx.js:9
hge @ main-DHJmLzGx.js:9
fge @ main-DHJmLzGx.js:9
Ige @ main-DHJmLzGx.js:9
R @ main-DHJmLzGx.js:2Understand this error
installHook.js:1 TypeError: e?.includes is not a function
    at kln (main-DHJmLzGx.js:585:39011)
    at e9 (main-DHJmLzGx.js:9:48989)
    at x9 (main-DHJmLzGx.js:9:72055)
    at Vme (main-DHJmLzGx.js:9:82551)
    at vge (main-DHJmLzGx.js:9:118645)
    at Bbt (main-DHJmLzGx.js:9:117664)
    at U9 (main-DHJmLzGx.js:9:117491)
    at fge (main-DHJmLzGx.js:9:114228)
    at jge (main-DHJmLzGx.js:9:126416)
    at f_ (main-DHJmLzGx.js:9:124942)
overrideMethod @ installHook.js:1
_me @ main-DHJmLzGx.js:9
Tme @ main-DHJmLzGx.js:9
Dme.Y.componentDidCatch.y.callback @ main-DHJmLzGx.js:9
Phe @ main-DHJmLzGx.js:9
$he @ main-DHJmLzGx.js:9
Qme @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
eg @ main-DHJmLzGx.js:9
Kme @ main-DHJmLzGx.js:9
kge @ main-DHJmLzGx.js:9
wge @ main-DHJmLzGx.js:9
hge @ main-DHJmLzGx.js:9
fge @ main-DHJmLzGx.js:9
jge @ main-DHJmLzGx.js:9
f_ @ main-DHJmLzGx.js:9
Oge @ main-DHJmLzGx.js:9
wge @ main-DHJmLzGx.js:9
hge @ main-DHJmLzGx.js:9
fge @ main-DHJmLzGx.js:9
Ige @ main-DHJmLzGx.js:9
R @ main-DHJmLzGx.js:2Understand this error
installHook.js:1 React Router caught the following error during render TypeError: e?.includes is not a function
    at kln (main-DHJmLzGx.js:585:39011)
    at e9 (main-DHJmLzGx.js:9:48989)
    at x9 (main-DHJmLzGx.js:9:72055)
    at Vme (main-DHJmLzGx.js:9:82551)
    at vge (main-DHJmLzGx.js:9:118645)
    at Bbt (main-DHJmLzGx.js:9:117664)
    at U9 (main-DHJmLzGx.js:9:117491)
    at fge (main-DHJmLzGx.js:9:114228)
    at jge (main-DHJmLzGx.js:9:126416)
    at f_ (main-DHJmLzGx.js:9:124942)
```





### Steps To Reproduce

1. Create a new Mastra project with `npm create mastra@latest`, cd into the folder and run the Studio locally with `npm run dev` or clone https://github.com/mikhael28/mastra-news-agent
2. If cloning the reproducible example, please add the relevant environment variables in `.env.example` and rename to `.env`
3. If you don't already have them, please create an API key for Browserbase and OpenAI particularly.
4. In the local Studio, navigate to the `Templates` tab, then select 'Browser Agent'
5. From here, add all required environment variables; double check that your Browserbase ProjectID and API key are not switched up.
6. Use OpenAI as your 'Template AI Model Provider' 
7. When picking 'Set AI Model for Template Installation' select OpenAI as the provider, and pick either 'codex-mini-latest' or 'gpt-5.1-codex' and click 'Install' once all fields are filled in.
8. View the crash details.

### Link to Minimal Reproducible Example

https://github.com/mikhael28/mastra-news-agent

### Expected Behavior

The template should install, thus generating the TypeScript code for the integration.

### Environment Information

```block
Running on Windows, with WSL2 running the Mastra Studio locally installed with npm.
LLM Provider - OpenAI
Browser - Chrome
 System:
    OS: Linux 6.6 Ubuntu 24.04.2 LTS 24.04.2 LTS (Noble Numbat)
    CPU: (20) x64 AMD Ryzen AI 9 365 w/ Radeon 880M
    Memory: 14.00 GB / 15.18 GB
    Container: Yes
    Shell: 5.2.21 - /bin/bash
  Binaries:
    Node: 22.16.0 - /home/mikhael/.nvm/versions/node/v22.16.0/bin/node
    npm: 11.4.1 - /home/mikhael/.nvm/versions/node/v22.16.0/bin/npm
    pnpm: 10.13.1 - /usr/local/bin/pnpm
    bun: 1.2.21 - /home/mikhael/.bun/bin/bun
  npmGlobalPackages:
    @expo/cli: 0.24.20
    corepack: 0.32.0
    mastra: 0.13.1
    mastracode: 0.6.0
    ngrok: 5.0.0-beta.2
    npm: 11.4.1
    serverless: 4.18.0
```

### Verification

- [x] I have searched the existing issues to make sure this is not a duplicate
- [x] I have included sufficient information for the team to reproduce and understand the issue

--- a/.changeset/easy-clubs-leave.md
+++ b/.changeset/easy-clubs-leave.md
@@ -1,0 +1,5 @@
+---
+'@mastra/playground-ui': patch
+---
+
+Fixed crash during template installation by ensuring error values from stream events are converted to strings before being passed to the UI. This prevents the 'e?.includes is not a function' TypeError when the server returns non-string error payloads.

--- a/.changeset/eighty-jobs-invite.md
+++ b/.changeset/eighty-jobs-invite.md
@@ -1,0 +1,5 @@
+---
+'@mastra/playground-ui': patch
+---
+
+Fixed crash when template installation errors are non-string values (e.g. objects). The error is now safely converted to a string before calling .includes(), preventing the 'e?.includes is not a function' TypeError in the studio.

--- a/packages/playground-ui/src/domains/templates/template-failure.tsx
+++ b/packages/playground-ui/src/domains/templates/template-failure.tsx
@@ -8,8 +8,10 @@ type TemplateFailureProps = {
 };
 
 export function TemplateFailure({ errorMsg, validationErrors }: TemplateFailureProps) {
-  const isSchemaError = errorMsg?.includes('Invalid schema for function');
-  const isValidationError = errorMsg?.includes('validation issue') || (validationErrors && validationErrors.length > 0);
+  const errorString = typeof errorMsg === 'string' ? errorMsg : errorMsg != null ? String(errorMsg) : undefined;
+  const isSchemaError = errorString?.includes('Invalid schema for function');
+  const isValidationError =
+    errorString?.includes('validation issue') || (validationErrors && validationErrors.length > 0);
 
   const getUserFriendlyMessage = () => {
     if (isValidationError) {
@@ -69,13 +71,13 @@ export function TemplateFailure({ errorMsg, validationErrors }: TemplateFailureP
       )}
 
       {/* General Error Details */}
-      {errorMsg && !isValidationError && (
+      {errorString && !isValidationError && (
         <details className="text-xs">
           <summary className="cursor-pointer text-neutral3 hover:text-neutral4 select-none text-center">
             Show Details
           </summary>
           <div className="mt-4 p-3 bg-gray-100 dark:bg-gray-800 rounded text-xs font-mono overflow-auto max-h-60 text-left">
-            <div className="whitespace-pre-wrap break-words">{errorMsg}</div>
+            <div className="whitespace-pre-wrap break-words">{errorString}</div>
           </div>
         </details>
       )}

--- a/packages/playground/src/hooks/use-templates.ts
+++ b/packages/playground/src/hooks/use-templates.ts
@@ -187,6 +187,20 @@ export const useGetTemplateInstallRun = () => {
   });
 };
 
+const normalizeError = (error: unknown): string => {
+  if (typeof error === 'string') return error;
+  if (error == null) return 'Unknown error';
+  if (error instanceof Error) return error.message;
+  if (typeof error === 'object') {
+    try {
+      return JSON.stringify(error);
+    } catch {
+      return String(error);
+    }
+  }
+  return String(error);
+};
+
 // Helper function to process template installation records (like workflows' sanitizeWorkflowWatchResult)
 const processTemplateInstallRecord = (
   record: { type: string; payload: any; runId?: string; eventTimestamp?: string },
@@ -312,14 +326,15 @@ const processTemplateInstallRecord = (
 
     // If this step failed, also set workflow-level error state
     if (status === 'failed' && hasError) {
+      const errorString = normalizeError(hasError);
       newState = {
         ...newState,
         status: 'failed',
-        error: hasError,
+        error: errorString,
         phase: 'error',
         failedStep: {
           id: stepId,
-          error: hasError,
+          error: errorString,
           description: record.payload.description || stepId,
         },
         payload: {
@@ -374,10 +389,11 @@ const processTemplateInstallRecord = (
   }
 
   if (record.type === 'error') {
+    const errorStr = normalizeError(record.payload.error);
     newState = {
       ...newState,
       status: 'failed',
-      error: record.payload.error,
+      error: errorStr,
       phase: 'error',
       payload: {
         ...newState.payload,

--- a/packages/playground/src/hooks/use-templates.ts
+++ b/packages/playground/src/hooks/use-templates.ts
@@ -190,7 +190,8 @@ export const useGetTemplateInstallRun = () => {
 const normalizeError = (error: unknown): string => {
   if (typeof error === 'string') return error;
   if (error == null) return 'Unknown error';
-  if (error instanceof Error) return error.message;
+  if (error instanceof Error)
+    return typeof error.message === 'string' && error.message.length > 0 ? error.message : 'Unknown error';
   if (typeof error === 'object') {
     try {
       return JSON.stringify(error);

--- a/packages/playground/src/pages/templates/template/index.tsx
+++ b/packages/playground/src/pages/templates/template/index.tsx
@@ -95,7 +95,8 @@ export default function Template() {
               setFailure(errorMessage);
               setCompletedRunValidationErrors(errors || []);
             } else {
-              setFailure(snapshot?.result?.message || snapshot?.result?.error || 'Template installation failed');
+              const errorValue = snapshot?.result?.message || snapshot?.result?.error || 'Template installation failed';
+              setFailure(typeof errorValue === 'string' ? errorValue : String(errorValue));
             }
           }
         })
@@ -239,7 +240,7 @@ export default function Template() {
     const result = streamResult || observeStreamResult;
 
     if (result?.phase === 'error' && result?.error) {
-      setFailure(result.error);
+      setFailure(typeof result.error === 'string' ? result.error : String(result.error));
     }
     // eslint-disable-next-line react-hooks/exhaustive-deps
   }, [streamResult?.phase, streamResult?.error, observeStreamResult?.phase, observeStreamResult?.error]);

